port/osx/omrosbacktrace_impl.c

```diff
@@ -0,0 +1,433 @@
+/*******************************************************************************
+ * Copyright (c) 2022, 2022 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+/**
+ * @file
+ * @ingroup Port
+ * @brief Stack backtracing support
+ */
+#ifndef _GNU_SOURCE
+#define _GNU_SOURCE
+#endif
+
+#include "omrport.h"
+#include "omrportpriv.h"
+#include "omrsignal_context.h"
+
+#include <dlfcn.h>
+#include <execinfo.h>
+#include <stdlib.h>
+#include <string.h>
+#define UNW_LOCAL_ONLY
+#include <libunwind.h>
+
+#include "omrintrospect.h"
+
+static uintptr_t protectedBacktrace(OMRPortLibrary *port, void *arg);
+static uintptr_t backtrace_sigprotect(OMRPortLibrary *portLibrary, J9PlatformThread *threadInfo, void **address_array, uintptr_t capacity);
+
+typedef struct frameData {
+	void **address_array;
+	uintptr_t capacity;
+} frameData;
+
+/**
+ * Handles segfaults where either an invalid address is called or a signal was received in kernel mode.
+ * Both cases leave the ip register address invalid and libunwind fails to iterate past _sigtramp. This
+ * function fetches previous frame information and restores correct register values to the unwind cursor.
+ * It also records the stack frame addresses of the remaining frames.
+ *
+ * @param[in] cursor keeps track of stack unwind state
+ * @param[out] array contains the backtrace addresses to be returned
+ * @param[in] size size of the array
+ * @param[in] nAddr number of addresses currently recorded in the array
+ *
+ * @return total number of addresses recorded in the array
+ */
+static uintptr_t
+handle_sigtramp(unw_cursor_t *cursor, void **array, uintptr_t size, uintptr_t nAddr)
+{
+	unw_word_t prevFrame = 0;
+	unw_word_t ip = 0;
+	ucontext_t *uc = NULL;
+	OMRUnixSignalInfo info;
+	const char *name = NULL;
+	void *reg = NULL;
+
+#if defined(OMR_ARCH_X86)
```

Comment:
Please move this `#if defined(OMR_ARCH_X86)` line to between `{` and `unw_word_t prevFrame = 0;` lines above.
Otherwise I get the following compiler errors on AArch64 macOS.
```
/Users/openj9/openj9-openjdk-jdk18/omr/port/osx/omrosbacktrace_impl.c:69:13: error: unused variable 'prevFrame' [-Werror,-Wunused-variable]
        unw_word_t prevFrame = 0;
                   ^
/Users/openj9/openj9-openjdk-jdk18/omr/port/osx/omrosbacktrace_impl.c:71:14: error: unused variable 'uc' [-Werror,-Wunused-variable]
        ucontext_t *uc = NULL;
                    ^
(four others omitted)
```